### PR TITLE
fix: properly format jwtkey prior to file save

### PR DESCRIPTION
**Pull request by peternhale**
@W-8856585@ Add function to format a jwt key before saving the key data to a file.

### CODE REVIEW

File: src/hubAuth.ts [hunk 1]
```diff
@@ -23,6 +23,38 @@ enum AuthStrategy {
...
+const formatJwtKey = (): string => {
+  if (env.getString('TESTKIT_JWT_KEY')) {
+    let keyLines = env.getString('TESTKIT_JWT_KEY', '').split('\n');
+    if (keyLines.length <= 1) {
+      const footer = '-----END RSA PRIVATE KEY-----';
+      const header = '-----BEGIN RSA PRIVATE KEY-----';
+      // strip out header and footer
+      const newKeyContents = env.getString('TESTKIT_JWT_KEY', '').replace(header, '').replace(footer, '');
+      // check to see if newlines were replaced with spaces
+      keyLines = newKeyContents.trim().split(/\s/);
```
Comment (peternhale) [inline src/hubAuth.ts:44]: On Macos, the newlines are replaced with spaces, so try a split on spaces. Not sure about other platforms.

I may change this to replace spaces with empty string to rid the code of the following conditionals.

File: src/hubAuth.ts [hunk 2]
```diff
@@ -23,6 +23,38 @@ enum AuthStrategy {
...
+      if (keyLines.length <= 1) {
+        // still one big string, split into 64 byte chucks
+        keyLines = [header, ...(newKeyContents.match(/.{1,64}/g) as string[]), footer];
```
Comment (peternhale) [inline src/hubAuth.ts:47]: Google was kind on this solution. Nice use of match and `{}` option

File: src/hubAuth.ts [hunk 3]
```diff
@@ -23,6 +23,38 @@ enum AuthStrategy {
...
+      } else {
+        keyLines = [header, ...keyLines, footer];
+      }
+    }
+    return keyLines.join('\n');
+  } else {
+    throw new Error('env var TESTKIT_JWT_KEY is undefined');
```
Comment (peternhale) [inline src/hubAuth.ts:54]: Throwing for now, since this function should only be called when one has a jwt key env var.

File: src/hubAuth.ts [hunk 4]
```diff
@@ -36,10 +68,10 @@ const DEFAULT_INSTANCE_URL = 'https://login.salesforce.com';
  */
 export const testkitHubAuth = (homeDir: string): void => {
   const logger = debug('testkit:authFromStubbedHome');
-  if (getAuthStrategy() === AuthStrategy.JWT && process.env.TESTKIT_JWT_KEY) {
+  if (getAuthStrategy() === AuthStrategy.JWT) {
     logger('trying jwt auth');
     const jwtKey = path.join(homeDir, 'jwtKey');
-    fs.writeFileSync(jwtKey, env.getString('TESTKIT_JWT_KEY', ''));
+    fs.writeFileSync(jwtKey, formatJwtKey());
```
Comment (peternhale) [inline src/hubAuth.ts:74]: Write a formatted key

File: src/hubAuth.ts [hunk 5]
```diff
@@ -50,7 +82,7 @@ export const testkitHubAuth = (homeDir: string): void => {
     );
     return;
   }
-  if (getAuthStrategy() === AuthStrategy.AUTH_URL && process.env.TESTKIT_AUTH_URL) {
+  if (getAuthStrategy() === AuthStrategy.AUTH_URL) {
```
Comment (peternhale) [inline src/hubAuth.ts:85]: `getAuthStrategy` uses the env var to determine result. No need to include the env var again here.

Comment (peternhale) [reply src/hubAuth.ts:85]: Replaced `process.env.` with `env.getString()` throughout.

File: src/hubAuth.ts [hunk 6]
```diff
@@ -23,6 +23,38 @@ enum AuthStrategy {
...
+ * reformatted so when saved to a file the RSA key file contents are formatted
+ * properly.
+ *
+ * Throws an error if function is called and the env var is undefined
+ *
+ * returns a string that complies with RSA private key file format
+ */
+const formatJwtKey = (): string => {
+  if (env.getString('TESTKIT_JWT_KEY')) {
+    let keyLines = env.getString('TESTKIT_JWT_KEY', '').split('\n');
```
Comment (mshanemc) [inline src/hubAuth.ts:37]: why default to an empty string?

[the remark this answers is not in the document]
Comment (mshanemc) [reply src/hubAuth.ts:37]: Would splitting by the OS-specific newline character be windows-friendlier?
 https://nodejs.org/api/os.html#os_os_eol

File: src/hubAuth.ts [hunk 7]
```diff
@@ -23,6 +23,38 @@ enum AuthStrategy {
...
+      // check to see if newlines were replaced with spaces
+      keyLines = newKeyContents.trim().split(/\s/);
+      if (keyLines.length <= 1) {
+        // still one big string, split into 64 byte chucks
+        keyLines = [header, ...(newKeyContents.match(/.{1,64}/g) as string[]), footer];
+      } else {
+        keyLines = [header, ...keyLines, footer];
+      }
+    }
+    return keyLines.join('\n');
```
Comment (mshanemc) [inline src/hubAuth.ts:52]: same comment (use os.eol)

**Review by mshanemc — APPROVED** [review]

**Review by shetzel — APPROVED** [review]
Approved.  2 comments:
1. Let's not forget about unit tests!
2. This smells more like a class to me than a collection of functions.  Unit testing might be tricky.

**Review by amphro — APPROVED** [review]

**Pull request merged**